ENH add feature_names_in_ in FeatureUnion

it176131 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
ENHANCEMENT #24754 
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
The `FeatureUnion` class did not previously have the `.feature_names_in_` attribute if fitted with a `pandas.DataFrame`. This allows access to the attribute.

#### Any other comments?
- modified:   sklearn/pipeline.py
  - added `self._check_feature_names(...)` to the `.fit(...)` method in `FeatureUnion` to allow access to the `.feature_names_in_` attribute if `X` has features names, e.g. a `pandas.DataFrame`
  - updated `FeatureUnion` docstring to reflect the addition of `.feature_names_in_` attribute

- modified:   sklearn/tests/test_pipeline.py
  - added `test_feature_union_feature_names_in_()` to test that `FeatureUnion` has a `.feature_names_in_` attribute if fitted with a `pandas.DataFrame` and not if fitted with a `numpy` array

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

it176131 committed:
modified:   sklearn/pipeline.py
	- added self._check_feature_names(...) to the .fit(...) method in FeatureUnion to allow access to the `.feature_names_in_` attribute if `X` has features names, e.g. a pandas.DataFrame
	- updated FeatureUnion docstring to reflect the addition of .feature_names_in_ attribute

modified:   sklearn/tests/test_pipeline.py
	- added test_feature_union_feature_names_in_() to test that FeatureUnion has a `.feature_names_in_` attribute if fitted with a pandas.DataFrame and not if fitted with a numpy array

it176131 committed:
modified:   doc/whats_new/v1.3.rst
	- changelog updated with description of work

Kshitij68 commented on doc/whats_new/v1.3.rst:
```diff
@@ -41,6 +41,10 @@ Changelog
   `Thomas Fan`_.
+- |Feature| :class:`pipeline.FeatureUnion` can now access the
+  `feature_names_in_` attribute if the value seen during `.fit` was a `pandas
+  .DataFrame`.
```
looking [here](https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/utils/validation.py#L1893), the check is not whether it is pandas dataframe, but whether an attribute `columns` is present. 

Maybe being more precise would help. (although most users tend to only use pandas/numpy with scikit-learn)

Kshitij68 replied:
or we could say what was written when this feature was introduced 
https://github.com/scikit-learn/scikit-learn/blame/main/doc/whats_new/v1.1.rst#L261

it176131 replied:
@Kshitij68 I've updated the changelog. What do you think?

thomasjpfan replied:
Since the change log is user facing and most people will be using a DataFrame, I prefer saying a pandas DataFrame. 

Implementation-wise we check for `columns`, because scikit-learn does not want a hard dependency on pandas. Although, technically we support "array-likes" with a `columns` attribute with all strings, the primary focus of the feature is for DataFrames.

it176131 replied:
> Since the change log is user facing and most people will be using a DataFrame, I prefer saying a pandas DataFrame.
> 
> Implementation-wise we check for `columns`, because scikit-learn does not want a hard dependency on pandas. Although, technically we support "array-likes" with a `columns` attribute with all strings, the primary focus of the feature is for DataFrames.

Should I revert back to the previous? Or add `pandas.DataFrame` to the description? Something like

```rst
- |Feature| :class:`pipeline.FeatureUnion` can now access the
  `feature_names_in_` attribute if the `X` value seen during `.fit` has a
  `columns` attribute and all columns are strings. e.g. when `X` is a 
  `pandas.DataFrame`.
  :pr:`25220` by :user:`Ian Thompson <ianiat11>`.
```

it176131 committed:
Merge remote-tracking branch 'origin/main' into feature_union_feature_names_in_

it176131 committed:
modified:   doc/whats_new/v1.3.rst
	- made changelog description more precise

it176131 committed:
modified:   doc/whats_new/v1.3.rst
	- typo -- removed period (.) before `columns`

thomasjpfan commented on sklearn/pipeline.py:
```diff
@@ -995,6 +995,12 @@ class FeatureUnion(TransformerMixin, _BaseComposition):
 
+    feature_names_in_ : ndarray of shape (`n_features_in_`,)
+        Names of features seen during :term:`fit`. Defined only when
+        `X` has feature names that are all strings.
```
This will end up to be inconsistent with when `n_features_in_` is defined. `n_features_in_` is only defined if the first transformer in `transformer_list` exposes it. So there are two paths forward:

1. Always define `n_features_in_` **and** `feature_names_in_`
2. Have `feature_names_in_` also depend on if `transformer_list` defines it.

Option 2 will require less discussion because that is the status quo. For example, `Pipeline` also exposes `feature_names_in_` based on the first `step`:

https://github.com/scikit-learn/scikit-learn/blob/fba66ded6c17023f3748d3cd2d49630fb7abbb28/sklearn/pipeline.py#L787-L791

So for this PR, I recommend going with option 2 since it is already an improvement.

it176131 replied:
If I add the following to `FeatureUnion`

```python
@property
def feature_names_in_(self):
    """Names of features seen during :term:`fit`."""
    # X is passed to all transformers so we just delegate to the first one
    return self.transformer_list[0][1].feature_names_in_
```

It works as expected. But if I instantiate `FeatureUnion` with a fitted transformer,

```python
scaler = StandardScaler()
scaler.fit(X_train)
union = FeatureUnion([("scale", scaler)])
assert hasattr(union, "feature_names_in_")
assert_array_equal(X_train.columns, union.feature_names_in_)
```

This will pass. Is this what we'd expect?

thomasjpfan replied:
Yea it is expected to pass, which is consistent with `Pipeline`. Your code snippet would be a good test to include in this PR.

Historically, `Pipeline` and `FeatureUnion` were developed before we were strict about scikit-learn's API conventions on "fitted attributes". There has been work to fix this: https://github.com/scikit-learn/scikit-learn/pull/8350, but it has been blocked by some real uses cases the current behavior enables.

it176131 replied:
Got it.

- [ ] remove `self._check_feature_names(X, reset=True)` from the `FeatureUnion`'s `.fit(...)` method
- [ ] add the `feature_names_in_` property to `FeatureUnion`
- [ ] update `test_feature_union_feature_names_in_()` to include a check when `FeatureUnion` is instantiated with a previously fitted transformer

thomasjpfan reviewed (commented):
Thanks for the PR!

it176131 committed:
Merge remote-tracking branch 'origin/main' into feature_union_feature_names_in_

it176131 committed:
modified:   sklearn/pipeline.py
	- removed `self._check_feature_names(...) from `.fit(...)` method in `FeatureUnion`
	- added `feature_names_in_()` property to `FeatureUnion` to use first transformer's `feature_names_in_` attribute if present

modified:   sklearn/tests/test_pipeline.py
	- updated docstring for `test_feature_union_feature_names_in_()` to be more precise
	- added additional assertions to check if the `feature_names_in_` attribute is available to `FeatureUnion` if it's instantiated with a transformer that has already been fit

it176131 committed:
modified:   doc/whats_new/v1.3.rst
	- updated changelog description to include `pandas.DataFrame`
	- corrected user signature to match github account

it176131 committed:
Merge branch 'main' into feature_union_feature_names_in_

it176131 committed:
modified:   sklearn/tests/test_pipeline.py
	- added pandas import to `test_feature_union_feature_names_in_` so ImportError in azure-pipelines will pass

it176131 committed:
Merge branch 'feature_union_feature_names_in_' of https://github.com/it176131/scikit-learn into feature_union_feature_names_in_

it176131 committed:
Merge remote-tracking branch 'origin/main' into feature_union_feature_names_in_

it176131 committed:
Merge remote-tracking branch 'origin/main' into feature_union_feature_names_in_

it176131 committed:
Merge remote-tracking branch 'origin/main' into feature_union_feature_names_in_

it176131 committed:
Merge remote-tracking branch 'origin/main' into feature_union_feature_names_in_

thomasjpfan commented on doc/whats_new/v1.3.rst:
```diff
@@ -91,6 +91,11 @@ Changelog
 - |Feature| :class:`pipeline.FeatureUnion` can now use indexing notation (e.g.
   `feature_union["scalar"]`) to access transformers by name. :pr:`25093` by
   `Thomas Fan`_.
+- |Feature| :class:`pipeline.FeatureUnion` can now access the
```
Nit:

```suggestion

- |Feature| :class:`pipeline.FeatureUnion` can now access the
```

thomasjpfan commented on sklearn/tests/test_pipeline.py:
```diff
@@ -1647,3 +1647,35 @@ def test_feature_union_getitem_error(key):
+    """Ensure feature union has `.feature_names_in_` attribute if `X` has a
+    `columns` attribute.
+
+    Test for #24754"""
```
Nit:

```suggestion
    Test for #24754.
    """
```

thomasjpfan commented on sklearn/tests/test_pipeline.py:
```diff
@@ -1647,3 +1647,35 @@ def test_feature_union_getitem_error(key):
+    pytest.importorskip("pandas")
+
+    X, _ = load_iris(as_frame=True, return_X_y=True)
+    X_train, X_test = train_test_split(X, random_state=0)
```
Since we are only testing the API for `feature_names_in_` and `X_test` is not used, we can use `X` directly in this test without `train_test_split`.

it176131 replied:
See commit [073daf7](https://github.com/scikit-learn/scikit-learn/pull/25220/commits/073daf7a37dfb69c79ba4fba692be5544a2cdaba)

thomasjpfan reviewed (commented):
Thank you for the update!

it176131 committed:
Update doc/whats_new/v1.3.rst

newline/whitespace between change log updates.

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

it176131 committed:
Update sklearn/tests/test_pipeline.py

added period at end of docstring

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

it176131 committed:
modified:   sklearn/tests/test_pipeline.py
	- removed train-test-split per code suggestion -- using `X` directly

it176131 committed:
Merge remote-tracking branch 'origin/main' into feature_union_feature_names_in_

it176131 committed:
Merge remote-tracking branch 'origin/main' into feature_union_feature_names_in_

it176131 committed:
Merge branch 'main' into feature_union_feature_names_in_

thomasjpfan reviewed (approved):
LGTM

it176131 committed:
Merge branch 'main' into feature_union_feature_names_in_

it176131 committed:
Merge branch 'main' into feature_union_feature_names_in_

it176131 committed:
Merge branch 'main' into feature_union_feature_names_in_

it176131 committed:
Merge branch 'main' into feature_union_feature_names_in_

it176131 committed:
Merge branch 'main' into feature_union_feature_names_in_

it176131 committed:
Merge branch 'main' into feature_union_feature_names_in_

it176131 committed:
Merge branch 'scikit-learn:main' into feature_union_feature_names_in_

it176131 committed:
Merge branch 'scikit-learn:main' into feature_union_feature_names_in_

it176131 committed:
Merge branch 'main' into feature_union_feature_names_in_

it176131 committed:
Merge branch 'main' into feature_union_feature_names_in_

it176131 committed:
Merge branch 'main' into feature_union_feature_names_in_

it176131 committed:
Merge branch 'main' into feature_union_feature_names_in_

it176131 committed:
Merge branch 'main' into feature_union_feature_names_in_

glemaitre committed:
Merge branch 'main' into feature_union_feature_names_in_

glemaitre reviewed (approved):
LGTM. Thanks @it176131
